File: src/crc.c
/*
 * Copyright (C) 2026 Andrii Shylenko
 * SPDX-License-Identifier: GPL-3.0-or-later
 *
 * This file is part of iolinki.
 * See LICENSE for details.
 */

/**
 * @file crc.c
 * @brief IO-Link message checksum implementation (A.1.6).
 * @ingroup iolinki_crc
 *
 * Implements the XOR-seeded, 8-to-6-bit compressed message checksum used to
 * protect every IO-Link M-sequence.
 */

#include "iolinki/crc.h"
#include "iolinki/utils.h"
#include <stddef.h>

/*
 * IO-Link message checksum, IO-Link Interface Specification V1.1.5 A.1.6.
 *
 * All octets (including the checksum/type octet with bits 0-5 zeroed) are XOR
 * processed with a seed of 0x52. The 8-bit result is compressed to 6 bits by
 * the equations in (A.1):
 *
 *   D5_6 = D7_8 xor D5_8 xor D3_8 xor D1_8
 *   D4_6 = D6_8 xor D4_8 xor D2_8 xor D0_8
 *   D3_6 = D7_8 xor D6_8
 *   D2_6 = D5_8 xor D4_8
 *   D1_6 = D3_8 xor D2_8
 *   D0_6 = D1_8 xor D0_8
 */
uint8_t iolink_checksum6(const uint8_t* octets, size_t len)
{
    uint8_t ck8 = 0x52U; /* A.1.6 seed value */

    if (!iolink_buf_is_valid(octets, len)) {
        return 0U;
    }

    for (size_t i = 0U; i < len; i++) {
        ck8 ^= octets[i];
    }

    const uint8_t b0 = (uint8_t) (ck8 & 0x01U);
    const uint8_t b1 = (uint8_t) ((ck8 >> 1U) & 0x01U);
    const uint8_t b2 = (uint8_t) ((ck8 >> 2U) & 0x01U);
    const uint8_t b3 = (uint8_t) ((ck8 >> 3U) & 0x01U);
    const uint8_t b4 = (uint8_t) ((ck8 >> 4U) & 0x01U);
    const uint8_t b5 = (uint8_t) ((ck8 >> 5U) & 0x01U);
    const uint8_t b6 = (uint8_t) ((ck8 >> 6U) & 0x01U);
    const uint8_t b7 = (uint8_t) ((ck8 >> 7U) & 0x01U);

    uint8_t ck6 = 0U;
    ck6 |= (uint8_t) ((b7 ^ b5 ^ b3 ^ b1) << 5U);
    ck6 |= (uint8_t) ((b6 ^ b4 ^ b2 ^ b0) << 4U);
    ck6 |= (uint8_t) ((b7 ^ b6) << 3U);
    ck6 |= (uint8_t) ((b5 ^ b4) << 2U);
    ck6 |= (uint8_t) ((b3 ^ b2) << 1U);
    ck6 |= (uint8_t) (b1 ^ b0);

    return ck6;
}

File: tools/virtual_master/virtual_master/crc.py
"""
Copyright (C) 2026 Andrii Shylenko
SPDX-License-Identifier: GPL-3.0-or-later

This file is part of iolinki.
See LICENSE for details.
"""

"""
Message checksum for IO-Link M-sequences.

Implements the A.1.6 message checksum (XOR with seed 0x52, compressed from
8 to 6 bits by equations (A.1)). Mirrors iolinki/src/crc.c.
"""

M_SEQUENCE_TYPE_MASK = 0xC0
CHECKSUM_MASK = 0x3F
CHECKSUM_SEED = 0x52


def checksum6(octets: bytes) -> int:
    """
    Calculate the IO-Link A.1.6 message checksum.

    Every octet is XOR processed with a seed of 0x52; the 8-bit result is
    compressed to 6 bits using the equations in (A.1). The checksum/type octet
    must be passed with its checksum bits (0-5) set to zero.

    Args:
        octets: Message octets with the checksum field zeroed.

    Returns:
        6-bit checksum value (0-63)
    """
    ck8 = CHECKSUM_SEED
    for octet in octets:
        ck8 ^= octet

    b = [(ck8 >> i) & 1 for i in range(8)]
    return (
        ((b[7] ^ b[5] ^ b[3] ^ b[1]) << 5)
        | ((b[6] ^ b[4] ^ b[2] ^ b[0]) << 4)
        | ((b[7] ^ b[6]) << 3)
        | ((b[5] ^ b[4]) << 2)
        | ((b[3] ^ b[2]) << 1)
        | (b[1] ^ b[0])
    )


def calculate_checksum_type0(mc: int, ckt: int = 0x00) -> int:
    """
    Calculate checksum for a Type 0 M-sequence (MC + CKT).

    Args:
        mc: Master Command byte
        ckt: Command/Key/Type byte; its checksum bits are zeroed

    Returns:
        6-bit checksum
    """
    return checksum6(bytes([mc, ckt & M_SEQUENCE_TYPE_MASK]))


def calculate_checksum_type1(mc: int, ckt: int, pd: bytes, od: int, od2: int = None) -> int:
    """
    Calculate checksum for a Type 1/2 M-sequence (MC + CKT + PD + OD).

    Args:
        mc: Master Command byte
        ckt: Command/Key/Type byte; its checksum bits are zeroed
        pd: Process Data bytes
        od: On-request Data byte (first byte)
        od2: Second OD byte (for Type 2 only, optional)

    Returns:
        6-bit checksum
    """
    data = bytes([mc, ckt & M_SEQUENCE_TYPE_MASK]) + pd + bytes([od])
    if od2 is not None:
        data += bytes([od2])
    return checksum6(data)


def verify_checksum(frame: bytes, expected_ck: int) -> bool:
    """
    Verify the checksum of a received frame.

    Args:
        frame: Frame data without the CK byte; the CKT/CKS octet (if present)
            must already have its checksum bits zeroed
        expected_ck: Received checksum octet (only bits 0-5 are compared)

    Returns:
        True if the checksum matches
    """
    return checksum6(frame) == (expected_ck & CHECKSUM_MASK)
